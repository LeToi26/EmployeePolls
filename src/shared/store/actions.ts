import { InitialState, Questions, User, Users } from "shared/model/poll";
import productApi from "../api/productApi";
import { Dispatch } from "@reduxjs/toolkit";

export const RECEIVE_USERS = "RECEIVE_USERS";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const LOGOUT_CURRENT_USER = "LOGOUT_CURRENT_USER";

export const CREATE_QUESTION = "CREATE_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export const ADD_ANSWER_USER = "ADD_ANSWER_USER";
export const CREATE_QUESTION_USER = "CREATE_QUESTION_USER";

export function initialData() {
  return (dispatch: Dispatch) => {
    return productApi
      .getInitialData()
      .then(({ users, questions }: { users: any; questions: any }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      });
  };
}

export function receiveUsers(users: Users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function receiveQuestions(questions: Questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function handleLogout() {
  return (dispatch: Dispatch) => {
    return dispatch(logoutCurrentUser());
  };
}

export function logoutCurrentUser() {
  return {
    type: LOGOUT_CURRENT_USER,
  };
}

export function handleLogin(username: string, password: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    const user: any = Object.values(State.users).find(
      (user: any) => user.id === username && user.password === password
    );
    if (!!user) {
      return dispatch(setCurrentUser(user));
    }
  };
}

export function setCurrentUser(authedUser: User | {}) {
  return {
    type: SET_CURRENT_USER,
    authedUser,
  };
}

export function handleAddAnswer(questionId: string, answer: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    return productApi
      .saveQuestionAnswer(State.authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addAnswerQuestion(State.authedUser.id, questionId, answer));
        dispatch(addAnswerUser(State.authedUser.id, questionId, answer));
      });
  };
}

function addAnswerQuestion(author: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function addAnswerUser(authedUser: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}

export function handleCreateQuestion(
  firstOption: string,
  secondOption: string
) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();

    return productApi
      .saveQuestion(firstOption, secondOption, State.authedUser)
      .then((question) => {
        dispatch(createQuestion(question));
        dispatch(createQuestionUser(question));
      });
  };
}

export function createQuestionUser(question: any) {
  const { author, id } = question;
  return {
    type: CREATE_QUESTION_USER,
    author,
    qid: id,
  };
}

function createQuestion(question: any) {
  return {
    type: CREATE_QUESTION,
    question,
  };
}
