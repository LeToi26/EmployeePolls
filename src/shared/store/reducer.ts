import { InitialState } from "shared/model/poll";
import {
  SET_CURRENT_USER,
  LOGOUT_CURRENT_USER,
  RECEIVE_USERS,
  RECEIVE_QUESTIONS,
  CREATE_QUESTION,
  CREATE_QUESTION_USER,
  ADD_ANSWER_USER,
  ADD_ANSWER_QUESTION,
} from "./actions";

const initialState: InitialState = {
  authedUser: null,
  users: {},
  questions: {},
};

export default function State(state = initialState, action: any) {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        authedUser: action.authedUser,
      };
    case LOGOUT_CURRENT_USER:
      return {
        ...state,
        authedUser: null,
      };
    case RECEIVE_USERS:
      return {
        ...state,
        users: action.users,
      };
    case RECEIVE_QUESTIONS:
      return {
        ...state,
        questions: action.questions,
      };
    case ADD_ANSWER_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.authedUser]: {
            ...state.users[action.authedUser as keyof typeof state],
            answers: {
              ...state.users[action.authedUser as keyof typeof state].answers,
              [action.qid]: action.answer,
            },
          },
        },
      };
    case ADD_ANSWER_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.qid]: {
            ...state.questions[action.qid as keyof typeof state],
            [action.answer]: {
              ...state.questions[action.qid as keyof typeof state][
                action.answer
              ],
              votes: state.questions[action.qid as keyof typeof state][
                action.answer
              ].votes.concat(action.author),
            },
          },
        },
      };
    case CREATE_QUESTION_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.author]: {
            ...state.users[action.author as keyof typeof state],
            questions: state.users[
              action.author as keyof typeof state
            ].questions.concat(action.qid),
          },
        },
      };
    case CREATE_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.question.id]: action.question,
        },
      };
    default:
      return state;
  }
}
