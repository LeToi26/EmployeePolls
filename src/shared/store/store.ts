import { configureStore } from "@reduxjs/toolkit";
import State from "./reducer";

export const store = configureStore({
  reducer: {
    State,
  },
});
