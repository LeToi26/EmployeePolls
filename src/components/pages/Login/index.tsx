import { FormEvent, useState } from "react";
import { connect } from "react-redux";
import { Navigate } from "react-router-dom";
import { InitialState } from "shared/model/poll";
import { handleLogin } from "shared/store/actions";

export interface LoginPropsEmployee {
  dispatch: any;
  authedUser: any;
}


export const LoginEmployee = ({ dispatch, authedUser }: LoginPropsEmployee) => {
  const [username, setUsername] = useState("sarahedo");
  const [password, setPassword] = useState("password123");

  if (authedUser) {
    const urlParams = new URLSearchParams(window.location.search);
    const previousUrl = urlParams.get("redirectTo");
    return <Navigate to={previousUrl ? previousUrl : "/"} />;
  }

  const onChangeUsernameStaff = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setUsername(value);
  };

  const onChangePasswordStaff = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setPassword(value);
  };

  const onSubmitData = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    dispatch(handleLogin(username, password));
    setUsername("");
    setPassword("");
  };

  return (
    <div className="flex min-h-full flex-col justify-center px-6 py-12 lg:px-8">
      <div className="sm:mx-auto sm:w-full sm:max-w-sm">
        <h2
          data-testid="login"
          className="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
        >
          Login
        </h2>
      </div>

      <div className="mt-10 sm:mx-auto sm:w-full sm:max-w-sm">
        <form className="space-y-6" onSubmit={onSubmitData}>
          <div>
            <label
              htmlFor="username"
              className="block text-sm font-medium leading-6 text-gray-900"
            >
              Username
            </label>
            <div className="mt-2">
              <input
                value={username}
                onChange={onChangeUsernameStaff}
                type="text"
                name="username"
                id="username"
                data-testid="username"
                autoComplete="username"
                required
                className="block w-full rounded-md border-0 px-1.5 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
              />
            </div>
          </div>

          <div>
            <div className="flex items-center justify-between">
              <label
                htmlFor="password"
                className="block text-sm font-medium leading-6 text-gray-900"
              >
                Password
              </label>
              <div className="text-sm"></div>
            </div>
            <div className="mt-2">
              <input
                value={password}
                onChange={onChangePasswordStaff}
                type="password"
                name="password"
                id="password"
                data-testid="password"
                autoComplete="current-password"
                required
                className="block w-full rounded-md border-0 px-1.5 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
              />
            </div>
          </div>

          <div>
            <button
              type="submit"
              data-testid="submit"
              className="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

const mapPropFromStates = ({ State }: { State: InitialState }) => ({
  authedUser: !!State.authedUser,
});

export default connect(mapPropFromStates)(LoginEmployee);

