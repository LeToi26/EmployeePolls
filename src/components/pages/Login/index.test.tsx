/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */

import { fireEvent, render, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "shared/store/store";
import Login from "./index";
import { initialData } from "shared/store/actions";
import { BrowserRouter } from "react-router-dom";

describe("Login", () => {
  it("should render the component", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("should clear user name and password after submit", async () => {
    await store.dispatch(initialData());

    const componnet = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const login = componnet.getByTestId("login");
    const username = componnet.getByTestId("username");
    const password = componnet.getByTestId("password");
    const submit = componnet.getByTestId("submit");
    expect(login).toBeInTheDocument();
    expect(username).toBeInTheDocument();
    expect(password).toBeInTheDocument();
    expect(submit).toBeInTheDocument();
    await fireEvent.change(username, { target: { value: "sarahedo" } });
    await fireEvent.change(password, {
      target: { value: "wrongpassword" },
    });
    await waitFor(() => {
      expect((username as HTMLInputElement).value).toBe("sarahedo");
    });
    await waitFor(() => {
      expect((password as HTMLInputElement).value).toBe("wrongpassword");
    });
    await fireEvent.click(submit);
    await waitFor(() => {
      expect(login).toBeInTheDocument();
    });
    expect((username as HTMLInputElement).value).toBe("");
    expect((password as HTMLInputElement).value).toBe("");
  });
});
