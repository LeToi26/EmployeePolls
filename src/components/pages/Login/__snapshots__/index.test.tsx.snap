// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex min-h-full flex-col justify-center px-6 py-12 lg:px-8"
      >
        <div
          class="sm:mx-auto sm:w-full sm:max-w-sm"
        >
          <h2
            class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
            data-testid="login"
          >
            Login
          </h2>
        </div>
        <div
          class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
        >
          <form
            class="space-y-6"
          >
            <div>
              <label
                class="block text-sm font-medium leading-6 text-gray-900"
                for="username"
              >
                Username
              </label>
              <div
                class="mt-2"
              >
                <input
                  autocomplete="username"
                  class="block w-full rounded-md border-0 px-1.5 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  data-testid="username"
                  id="username"
                  name="username"
                  required=""
                  type="text"
                  value="sarahedo"
                />
              </div>
            </div>
            <div>
              <div
                class="flex items-center justify-between"
              >
                <label
                  class="block text-sm font-medium leading-6 text-gray-900"
                  for="password"
                >
                  Password
                </label>
                <div
                  class="text-sm"
                />
              </div>
              <div
                class="mt-2"
              >
                <input
                  autocomplete="current-password"
                  class="block w-full rounded-md border-0 px-1.5 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  data-testid="password"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value="password123"
                />
              </div>
            </div>
            <div>
              <button
                class="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
                data-testid="submit"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex min-h-full flex-col justify-center px-6 py-12 lg:px-8"
    >
      <div
        class="sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <h2
          class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
          data-testid="login"
        >
          Login
        </h2>
      </div>
      <div
        class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <form
          class="space-y-6"
        >
          <div>
            <label
              class="block text-sm font-medium leading-6 text-gray-900"
              for="username"
            >
              Username
            </label>
            <div
              class="mt-2"
            >
              <input
                autocomplete="username"
                class="block w-full rounded-md border-0 px-1.5 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                data-testid="username"
                id="username"
                name="username"
                required=""
                type="text"
                value="sarahedo"
              />
            </div>
          </div>
          <div>
            <div
              class="flex items-center justify-between"
            >
              <label
                class="block text-sm font-medium leading-6 text-gray-900"
                for="password"
              >
                Password
              </label>
              <div
                class="text-sm"
              />
            </div>
            <div
              class="mt-2"
            >
              <input
                autocomplete="current-password"
                class="block w-full rounded-md border-0 px-1.5 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                data-testid="password"
                id="password"
                name="password"
                required=""
                type="password"
                value="password123"
              />
            </div>
          </div>
          <div>
            <button
              class="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
              data-testid="submit"
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
